File: rust_test_crate/tests/15_system_return_chain.rs

// Tests system.return in enter handlers (Rust version)
// Note: Rust uses native return pattern, chain semantics differ from Python/TS


use std::collections::HashMap;

#[derive(Clone, Debug)]
struct SystemReturnChainTestFrameEvent {
    message: String,
}

impl SystemReturnChainTestFrameEvent {
    fn new(message: &str) -> Self {
        Self { message: message.to_string() }
    }
}

#[derive(Clone)]
struct SystemReturnChainTestCompartment {
    state: String,
    state_vars: std::collections::HashMap<String, i32>,
    forward_event: Option<SystemReturnChainTestFrameEvent>,
}

impl SystemReturnChainTestCompartment {
    fn new(state: &str) -> Self {
        Self {
            state: state.to_string(),
            state_vars: std::collections::HashMap::new(),
            forward_event: None,
        }
    }
}

#[derive(Clone)]
enum SystemReturnChainTestStateContext {
    Start,
    EnterSetter,
    BothSet,
    Empty,
}

impl Default for SystemReturnChainTestStateContext {
    fn default() -> Self {
        SystemReturnChainTestStateContext::Start
    }
}

pub struct SystemReturnChainTest {
    _state_stack: Vec<(String, SystemReturnChainTestStateContext)>,
    __compartment: SystemReturnChainTestCompartment,
    __next_compartment: Option<SystemReturnChainTestCompartment>,
}

impl SystemReturnChainTest {
    pub fn new() -> Self {
        let mut this = Self {
            _state_stack: vec![],
            __compartment: SystemReturnChainTestCompartment::new("Start"),
            __next_compartment: None,
        };
let __frame_event = SystemReturnChainTestFrameEvent::new("$>");
this.__kernel(__frame_event);
        this
    }

    fn __kernel(&mut self, __e: SystemReturnChainTestFrameEvent) {
// Route event to current state
self.__router(&__e);
// Process any pending transition
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    // Exit current state
    let exit_event = SystemReturnChainTestFrameEvent::new("$<");
    self.__router(&exit_event);
    // Switch to new compartment
    self.__compartment = next_compartment;
    // Enter new state (or forward event)
    if self.__compartment.forward_event.is_none() {
        let enter_event = SystemReturnChainTestFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        // Forward event to new state
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            // Forwarding enter event - just send it
            self.__router(&forward_event);
        } else {
            // Forwarding other event - send $> first, then forward
            let enter_event = SystemReturnChainTestFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
    }

    fn __router(&mut self, __e: &SystemReturnChainTestFrameEvent) {
match self.__compartment.state.as_str() {
    "Start" => self._state_Start(__e),
    "EnterSetter" => self._state_EnterSetter(__e),
    "BothSet" => self._state_BothSet(__e),
    _ => {}
}
    }

    fn __transition(&mut self, next_compartment: SystemReturnChainTestCompartment) {
self.__next_compartment = Some(next_compartment);
    }

    fn _state_stack_push(&mut self) {
let state_context = match self.__compartment.state.as_str() {
    "Start" => SystemReturnChainTestStateContext::Start,
    "EnterSetter" => SystemReturnChainTestStateContext::EnterSetter,
    "BothSet" => SystemReturnChainTestStateContext::BothSet,
    _ => SystemReturnChainTestStateContext::Empty,
};
self._state_stack.push((self.__compartment.state.clone(), state_context));
    }

    fn _state_stack_pop(&mut self) {
let (saved_state, state_context) = self._state_stack.pop().unwrap();
let exit_event = SystemReturnChainTestFrameEvent::new("$<");
self.__router(&exit_event);
self.__compartment.state = saved_state;
match state_context {
    SystemReturnChainTestStateContext::Start => {}
    SystemReturnChainTestStateContext::EnterSetter => {}
    SystemReturnChainTestStateContext::BothSet => {}
    SystemReturnChainTestStateContext::Empty => {}
}
    }

    pub fn get_state_num(&mut self) -> i32 {
let __e = SystemReturnChainTestFrameEvent::new("get_state_num");
match self.__compartment.state.as_str() {
            "Start" => self._s_Start_get_state_num(&__e),
            "EnterSetter" => self._s_EnterSetter_get_state_num(&__e),
            "BothSet" => self._s_BothSet_get_state_num(&__e),
            _ => Default::default(),
        }
    }

    fn _state_EnterSetter(&mut self, __e: &SystemReturnChainTestFrameEvent) {
match __e.message.as_str() {
    "get_state_num" => { self._s_EnterSetter_get_state_num(__e); }
    _ => {}
}
    }

    fn _state_BothSet(&mut self, __e: &SystemReturnChainTestFrameEvent) {
match __e.message.as_str() {
    "get_state_num" => { self._s_BothSet_get_state_num(__e); }
    _ => {}
}
    }

    fn _state_Start(&mut self, __e: &SystemReturnChainTestFrameEvent) {
match __e.message.as_str() {
    "get_state_num" => { self._s_Start_get_state_num(__e); }
    _ => {}
}
    }

    fn _s_EnterSetter_get_state_num(&mut self, __e: &SystemReturnChainTestFrameEvent) -> i32 {
return 2;
    }

    fn _s_BothSet_get_state_num(&mut self, __e: &SystemReturnChainTestFrameEvent) -> i32 {
return 3;
    }

    fn _s_Start_get_state_num(&mut self, __e: &SystemReturnChainTestFrameEvent) -> i32 {
return 1;
    }
}


fn main() {
    println!("=== Test 15: System Return (Rust) ===");

    let mut s1 = SystemReturnChainTest::new();
    let state = s1.get_state_num();
    assert_eq!(state, 1, "Expected state 1, got {}", state);
    println!("1. get_state_num() = {}", state);

    println!("PASS: System return works correctly");
}

File: output/cpp/tests/15_system_return_chain.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>
#include <functional>


#include <iostream>
#include <string>
#include <cassert>

// Tests that @@:return follows "last writer wins" across transition lifecycle

class SystemReturnChainTestFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    SystemReturnChainTestFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class SystemReturnChainTestFrameContext {
public:
    SystemReturnChainTestFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    SystemReturnChainTestFrameContext(SystemReturnChainTestFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class SystemReturnChainTestCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<SystemReturnChainTestFrameEvent> forward_event;
    std::unique_ptr<SystemReturnChainTestCompartment> parent_compartment;

    explicit SystemReturnChainTestCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<SystemReturnChainTestCompartment> clone() const {
        auto c = std::make_unique<SystemReturnChainTestCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class SystemReturnChainTest {
private:
    std::vector<std::unique_ptr<SystemReturnChainTestCompartment>> _state_stack;
    std::unique_ptr<SystemReturnChainTestCompartment> __compartment;
    std::unique_ptr<SystemReturnChainTestCompartment> __next_compartment;
    std::vector<SystemReturnChainTestFrameContext> _context_stack;

    void __kernel(SystemReturnChainTestFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            SystemReturnChainTestFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                SystemReturnChainTestFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    SystemReturnChainTestFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(SystemReturnChainTestFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Start") {
            _state_Start(__e);
        } else if (state_name == "EnterSetter") {
            _state_EnterSetter(__e);
        } else if (state_name == "BothSet") {
            _state_BothSet(__e);
        }
    }

    void __transition(std::unique_ptr<SystemReturnChainTestCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_Start(SystemReturnChainTestFrameEvent& __e) {
        if (__e._message == "<$") {
            _context_stack.back()._return = std::any(std::string("from_exit"));
        } else if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("Start"));
            return;;
        } else if (__e._message == "test_enter_sets") {
            auto __new_compartment = std::make_unique<SystemReturnChainTestCompartment>("EnterSetter");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        } else if (__e._message == "test_exit_then_enter") {
            auto __new_compartment = std::make_unique<SystemReturnChainTestCompartment>("BothSet");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        }
    }

    void _state_EnterSetter(SystemReturnChainTestFrameEvent& __e) {
        if (__e._message == "$>") {
            _context_stack.back()._return = std::any(std::string("from_enter"));
        } else if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("EnterSetter"));
            return;;
        }
    }

    void _state_BothSet(SystemReturnChainTestFrameEvent& __e) {
        if (__e._message == "$>") {
            _context_stack.back()._return = std::any(std::string("enter_wins"));
        } else if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("BothSet"));
            return;;
        }
    }

public:
    SystemReturnChainTest() {
        __compartment = std::make_unique<SystemReturnChainTestCompartment>("Start");
        SystemReturnChainTestFrameEvent __frame_event("$>");
        SystemReturnChainTestFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    std::string test_enter_sets() {
        SystemReturnChainTestFrameEvent __e("test_enter_sets");
        SystemReturnChainTestFrameContext __ctx(std::move(__e), std::any(std::string()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::string>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    std::string test_exit_then_enter() {
        SystemReturnChainTestFrameEvent __e("test_exit_then_enter");
        SystemReturnChainTestFrameContext __ctx(std::move(__e), std::any(std::string()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::string>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    std::string get_state() {
        SystemReturnChainTestFrameEvent __e("get_state");
        SystemReturnChainTestFrameContext __ctx(std::move(__e), std::any(std::string()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::string>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }
};

int main() {
    printf("=== Test 15: System Return Chain (Last Writer Wins) ===\n");

    SystemReturnChainTest s1;
    std::string result1 = s1.test_enter_sets();
    if (result1 != "from_enter") {
        printf("FAIL: Expected 'from_enter', got '%s'\n", result1.c_str());
        assert(false);
    }
    if (s1.get_state() != "EnterSetter") {
        printf("FAIL: Expected state 'EnterSetter'\n");
        assert(false);
    }
    printf("1. Exit set then enter set - enter wins: '%s'\n", result1.c_str());

    SystemReturnChainTest s2;
    std::string result2 = s2.test_exit_then_enter();
    if (result2 != "enter_wins") {
        printf("FAIL: Expected 'enter_wins', got '%s'\n", result2.c_str());
        assert(false);
    }
    if (s2.get_state() != "BothSet") {
        printf("FAIL: Expected state 'BothSet'\n");
        assert(false);
    }
    printf("2. Both set - enter wins: '%s'\n", result2.c_str());

    printf("PASS: System return chain (last writer wins) works correctly\n");
    return 0;
}
